File: test_cafekiosk/src/main/java/com/testcafekiosk/spring/api/service/ProductService.java
//package com.testcafekiosk.spring.api.service;
//
//import com.testcafekiosk.spring.api.service.product.response.ProductResponse;
//import com.testcafekiosk.spring.domain.product.Product;
//import com.testcafekiosk.spring.domain.product.ProductRepository;
//import com.testcafekiosk.spring.domain.product.ProductSellingType;
//import lombok.RequiredArgsConstructor;
//import org.springframework.stereotype.Service;
//
//import java.util.List;
//import java.util.stream.Collectors;
//
//@Service
//@RequiredArgsConstructor
//public class ProductService {
//
//    private final ProductRepository productRepository;
//
//    public List<ProductResponse> getSellingProducts() {
//        List<Product> products = productRepository.findAllBySellingStatusIn(ProductSellingType.forDisplay());
//
//        return products.stream()
//                .map(ProductResponse::of)
//                .collect(Collectors.toList());
//    }
//}

File: test_cafekiosk/src/main/kotlin/com/testcafekiosk/spring/api/service/product/ProductService.kt
package com.testcafekiosk.spring.api.service.product

import com.testcafekiosk.spring.api.controller.product.request.ProductCreateRequest
import com.testcafekiosk.spring.api.service.product.response.ProductResponse
import com.testcafekiosk.spring.domain.product.ProductRepository
import com.testcafekiosk.spring.domain.product.ProductSellingStatus
import org.springframework.stereotype.Service

@Service
class ProductService(
    private val productRepository: ProductRepository,
) {

    fun getSellingProducts(): List<ProductResponse> {
        val products = productRepository.findAllBySellingStatusIn(ProductSellingStatus.forDisplay())

        return products
            .map { ProductResponse.from(it) }
            .toList()
    }

    fun createProduct(request: ProductCreateRequest): ProductResponse {
        val nextProductNumber = createNextProductNumber()

        var product = request.toEntity(nextProductNumber)
        product = productRepository.save(product)

        return ProductResponse.from(product)
    }

    private fun createNextProductNumber(): String {
        val latestProductNumber = productRepository.findLatestProductNumber()
            ?: return "001"

        val latestProductNumberInt = latestProductNumber.toInt()
        val nextProductNumberInt = latestProductNumberInt + 1

        return String.format("%03d", nextProductNumberInt)
    }
}